design/automatic-network-configuration.md

```diff
@@ -0,0 +1,738 @@
+    deleted, and set the state of CR to `Idle`.
+- Idle - Steady state
+    1. Indicates waiting for spec.configurationRef to be assigned.
+    2. The state handler check spec.configurationRef's value,
+    if isn't nil set the state of CR to Configuring
+    3. If deletionTimestamp set, goes to `Deleting`.
+- Configuring
+    1. Indicates that the port is being configured.
+    2. The state handler configures the port's network and checks configuration
+    progress. If finished set the state of CR to `Active` state.
+- Active - Steady State
+    1. Indicates that the port configuration is complete.
+    2. The state handler check whether the target configuration is consistent
+    with the actual configuration, return to `Configuring` state and clean
+    `status.configurationRef` when inconsistent.
+    3. If deletionTimestamp is set -> go to `Cleaning`
+- Cleaning
+    1. Indicates that the port configuration is being cleared.
+    2. The state handler deconfigure port's network and check deconfiguration
+    progress, when finished clean `spec.configurationRef` and
+    `status.configurationRef` then set CR's state to `Idle` state.
+    3. If deletionTimestamp set, goes to Deleting.
+- Deleting Reached when in Idle state + deletionTimestamp set
+    1. Indicates that the port configuration has been cleared.
+    2. Prepares for deletion.
+    3. The state handler will remove finalizers.
+
+![](./images/automatic-network-configuration-state.png)
+
+#### SmartNIC
+
+##### SmartNICPort CRD
+
+```yaml
+apiVersion: v1alpha1
+kind: SmartNicPort
+metaData:
+  name: port2
+  ownerRef:
+    name: smart1
+  finalizers:
+  - defaulf-bm0
+spec:
+  configurationRef:
+    name: sn1
+    namespace: default
+  # Represents the next port information of this port link
+  nextPortRef:
+    name: port3
+    kind: SwitchPort
+    namespace: default
+status:
+  # Indicates the actual configuration status of the port
+  state: Configured
+  # Indicates the configuration information currently applied to the port
+  configurationRef:
+    name: sn1
+    namespace: default
+```
+
+##### SmartNicConfiguration CRD
+
+```yaml
+apiVersion: v1alpha1
+kind: SmartNicConfiguration
+metaData:
+  name: sc1
+  namespace: default
+  finalizers:
+    - default-port0
+spec:
+  ...
+  nextConfigurationRef:
+  name: sp1
+  kind: SwitchPortConfiguration
+  namespace: default
+```
+
+### Multi-tenancy models
+
+- No multi-tenancy : one set of controllers deployed, all objects are
+  reconciled by the same controllers.
+- "Limited" multi-tenancy : one set of CRs (the most up-to-date), one set of
+  webhooks and a single network controller with clusterRole and in its own
+  namespace. All other controllers (CAPI + CAPM3) per namespace. Permission
+  separation between namespaces.
+- Full multi-tenancy : one set of CRs (the most up-to-date), one set of
+  webhooks. All other controllers per namespace. No permission separation
+  within the same namespace.
```

Comment:
I would not worry so much about running multiple controllers for different namespaces (this is never going to work well for k8s due to the CRD and webhooks being global, as you've identified).

What is more interesting is making sure that roles can be separate. So e.g. an administrator can assign a given Host to one of several tenants (each with their own namespace) and allow the tenants to control how it is provisioned (e.g. by using CAPM3, or something else) while making sure that e.g. only the administrator can define which networks the Host can connect to.

This will only actually be possible once the BaremetalHost itself is decomposed in some way along the lines of my draft proposal in https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit?usp=sharing (you must be a member of the Metal³ Google Group to view this doc), but it is worth thinking about how the network configuration would work in such a scenario. In particular, I think this scenario suggests that the network config should in future be associated with a BaremetalAllocation, not a BaremetalImage (i.e. deployment), and it follows from that perhaps the network configuration should precede the Host being selected by the CAPM3 rather than being part of the provisioning done by CAPM3.